src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java

```diff
@@ -1263,6 +1284,209 @@ public void visitMethodDef(JCMethodDecl tree) {
+            super.visitLambda(lambda);
+            classDeclAndLambdaHelper(TreeInfo.symbolsFor(lambda.body));
+        }
+
+        @Override
+        public void visitClassDef(JCClassDecl classDecl) {
+            super.visitClassDef(classDecl);
+            classDeclAndLambdaHelper(TreeInfo.symbolsFor(classDecl.defs));
+        }
+
+        private void classDeclAndLambdaHelper(java.util.List<TreeInfo.SymAndTree> symbols) {
+            for (TreeInfo.SymAndTree symAndTree : symbols) {
+                Symbol sym = symAndTree.symbol();
+                JCTree tree = TreeInfo.skipParens(symAndTree.tree());
+                if (sym.kind == VAR &&
+                        rs.isEarlyReference(
+                                true, // localEnv could potentially have the `ctorPrologue` field set to false
+                                localEnv,
+                                tree instanceof JCFieldAccess fa ?
+                                        fa.selected :
+                                        null,
+                                sym)) {
+                    reportPrologueError(tree, sym);
+                }
+            }
+        }
+
+        private void reportPrologueError(JCTree tree, Symbol sym) {
+            preview.checkSourceLevel(tree, Feature.FLEXIBLE_CONSTRUCTORS);
+            log.error(tree, Errors.CantRefBeforeCtorCalled(sym));
+        }
+
+        @Override
+        public void visitApply(JCMethodInvocation tree) {
+            super.visitApply(tree);
+            Name name = TreeInfo.name(tree.meth);
+            boolean isConstructorCall = name == names._this || name == names._super;
+            Symbol msym = TreeInfo.symbolFor(tree.meth);
+            // is this an instance method call or an illegal constructor invocation like: `this.super()`?
+            if (msym != null && // for erroneous invocations msym can be null, ignore those
+                (!isConstructorCall ||
+                isConstructorCall && tree.meth.hasTag(SELECT))) {
+                if (rs.isEarlyReference(
+                        true,
+                        localEnv,
+                        tree.meth instanceof JCFieldAccess fa ? fa.selected : null,
+                        msym))
+                    reportPrologueError(tree.meth, msym);
+            }
+        }
+
+        @Override
+        public void visitIdent(JCIdent tree) {
+            // skip if this identifier is part of a select, context is important here
+            if (!analyzingSelect) {
+                analyzeSymbol(tree);
+            }
+        }
+
+        boolean analyzingSelect = false;
+
+        @Override
+        public void visitSelect(JCFieldAccess tree) {
+            // skip if part of a larger select, context is important here
+            if (!analyzingSelect) {
+                boolean previousAnalyzingSelect = analyzingSelect;
+                try {
+                    analyzingSelect = true;
+                    super.visitSelect(tree);
```

Comment:
Can't we cut recursion here (instead of using `analyzingSelect` ? That's also what the new `TreeInfo.symbolsFor` does. In general it seems like these two visitors are trying to do similar things but are not 100% aligned?

Reply:
(e.g. we really need to make sure that `analyzeSelect` is not applied too broadly)

Reply:
yep :(